[gsi_web] Adds Sign In button.

ditman opened the pull request:
* Adds web-only renderButton method.
* Implements new platform interface methods.
  * `canAccessScopes` method and `userDataEvents` stream.
* Adds button_tester example app.
* Adds a flex view that reflows flutter apps from its contents.

**Breaking change**: `silentSignIn` now returns a non-null user, after popular demand (and adding a method to check if a user `canAccessScopes`).

## Issues

* Needs: https://github.com/flutter/packages/pull/3635
* Continues landing: https://github.com/flutter/packages/pull/3461
* Enables fixing: https://github.com/flutter/flutter/issues/121781
  * Enables fixing: https://github.com/flutter/flutter/issues/121275

## Testing

* Button Test App deployed here: https://dit-gis-button-test.web.app
 
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

yjbanov commented on packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart:
```diff
@@ -110,29 +147,66 @@ class GoogleSignInPlugin extends GoogleSignInPlatform {
-    // However, if this succeeds, we'll save a People API request later.
-    return _gisClient.signInSilently().then((_) => null);
+    // The new user is being injected from the `userDataEvents` Stream.
+    return _gisClient!.signInSilently(); //.then((_) => null);
```
nit: dead code starting with `//.then(...`

ditman replied:
Good eye, removing these across the package.

yjbanov commented on packages/google_sign_in/google_sign_in_web/lib/src/button_configuration.dart:
```diff
@@ -0,0 +1,202 @@
+
+/// A class to configure the Google Sign-In Button for web.
+///
+/// https://developers.google.com/identity/gsi/web/reference/js-reference#GsiButtonConfiguration
```
nit: Flutter style typically wants something like this:

```
/// See also:
///   * link
```

This comment applies throughout the file.

ditman replied:
Done.

yjbanov commented on packages/google_sign_in/google_sign_in_web/lib/src/button_configuration.dart:
```diff
@@ -0,0 +1,202 @@
+
+  /// The circle-shaped button.
+  pill;
+  // Does this need circle and square?
```
Maybe make this a github issue?

ditman replied:
I'll create an internal issue, this is a Google designed API.

ditman replied:
Opened b/276957876

yjbanov commented on packages/google_sign_in/google_sign_in_web/lib/src/flexible_size_html_element_view.dart:
```diff
@@ -0,0 +1,140 @@
+  @override
+  Widget build(BuildContext context) {
+    return SizedBox.fromSize(
+      size: _lastReportedSize ?? widget.initialSize ?? const Size(1, 1),
```
FYI: using `BoxConstraints` from a `LayoutBuilder` may provide a better first guess than `Size(1, 1)`.

ditman replied:
For the best user experience in this usage, the `initialSize` should be guesstimated off of the button configuration.

The layout builder however tells me the constraints that flutter will apply to my widget, but it does not tell me what's the rendered size of the contents of the html element view, so I'm not sure that the estimation is going to be better (if I understand the layout builder correctly, in that case the widget will start large, and then likely shrink, and with Size(1,1) it'll start small and likely grow?)

yjbanov reviewed (approved):
![lgtm](https://user-images.githubusercontent.com/211513/32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d.png)

ditman commented on packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart:
```diff
@@ -28,23 +45,45 @@ class GoogleSignInPlugin extends GoogleSignInPlatform {
   /// background.
   ///
   /// The plugin is completely initialized when [initialized] completed.
```
I kept the `@visibleForTesting` (otherwise it should be _private), and improved the documentation slightly.

ditman commented:
Hey, @stuartmorgan just a quick heads-up that this is almost ready to land. Please do take a look!

I've updated the README with the latest changes; instructing users to `canAccessScopes` after sign in on the web, and removing the mentions that `signInSilently` always returns `null` (because it doesn't anymore).

(The latter change made me want to release this as a major version change, so people who have already migrated to 0.11 get these changes by accident)

I've also added a new test app (that only uses this plugin) to test what the button looks like with different configuration values (I've deployed the latest version to the URL I linked in the description).

Thanks! 🙏

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/README.md:
```diff
@@ -49,48 +46,73 @@ See more differences in the following migration guides:
+check if a user is Authorized or not.
+
+It is also possible that Authorizations expire while users are using an App
+(after 3600 seconds), so Apps should monitor response failures from the APIs, and
```
Nit: apps

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/README.md:
```diff
@@ -49,48 +46,73 @@ See more differences in the following migration guides:
+[OpenID Connect scopes](https://developers.google.com/identity/protocols/oauth2/scopes#openid-connect).
 
-#### `signInSilently` always returns `null`
+If an App only needs an `idToken`, or the OpenID Connect scopes, the Authentication
```
app

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/README.md:
```diff
@@ -49,48 +46,73 @@ See more differences in the following migration guides:
-The information retrieved from the People API is used to complete data for the
-[`GoogleSignInAccount`](https://pub.dev/documentation/google_sign_in/latest/google_sign_in/GoogleSignInAccount-class.html)
-object that is returned after `signIn` completes successfully.
+Only if the scopes required by an App are different from the
```
app

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/README.md:
```diff
@@ -49,48 +46,73 @@ See more differences in the following migration guides:
+longer feasible on the web.
 
-Only after `signIn` or `requestScopes`, a user will be fully formed.
+The web plugin attempts to simulate the old `signIn` behavior by using the 
+[Oauth Implicit pop-up flow](https://developers.google.com/identity/oauth2/web/guides/use-token-model), which authenticates and authorizes users.
```
OAuth

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/README.md:
```diff
@@ -49,48 +46,73 @@ See more differences in the following migration guides:
-The GIS-backed plugin always returns `null` from `signInSilently`, to force apps
-that expect the former logic to perform a full `signIn`, which will result in a
-fully Authenticated and Authorized user, and making this migration easier.
+The drawback of this approach is that the Oauth flow **only returns an `accessToken`**,
```
OAuth

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/README.md:
```diff
@@ -49,48 +46,73 @@ See more differences in the following migration guides:
+users, and then `canAccessScopes` and `requestScopes` to authorize the `scopes`
+that are needed.
+
+#### Why is the `idToken` `null` after `signIn`
```
Nit: missing ?

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/example/lib/button_tester.dart:
```diff
@@ -0,0 +1,103 @@
+
+// The instance of the plugin is automatically created by Flutter before calling
+// our main code, let's grab it directly from the Platform interface of the plugin.
+GoogleSignInPlugin _plugin =
```
`final`?

ditman replied:
Yes!

stuartmorgan commented on packages/google_sign_in/google_sign_in_web/lib/google_sign_in_web.dart:
```diff
@@ -27,24 +44,56 @@ class GoogleSignInPlugin extends GoogleSignInPlatform {
+  // A future that completes when the `init` call is done.
+  final Completer<void> _initDone = Completer<void>();
+  // A (synchronous) marker to assert that `init` has been called.
   bool _isInitCalled = false;
```
Optional: would it make sense to remove this and make `_initDone` optional, with a null future indicating that it hasn't been called, so there's only one thing tracking the init flow?

ditman replied:
Actually, yes, it works Just Fine. Thanks for the suggestion!

stuartmorgan reviewed (approved):
The metadata LGTM with nits; I mostly didn't review the implementation part since that was already done.

ditman commented:
Let's go with this! When programmers attempt to use `signIn` in the web, the plugin will warn them that there's better alternatives, and point them to the docs!
